# Урок 5. Проверка статистических гипотез. P-значения. Доверительные интервалы. A/B-тестирование
https://matica.org.ua/primery/primery/primery-resheniia-tipovykh-zadach-po-matematicheskoi-statistike

Задача 1

Известно, что генеральная совокупность распределена нормально с известным средним квадратическим отклонением 16. Найти доверительный интервал для оценки математического ожидания mu с надёжностью 0.95, если выборочное среднее равно 80, а объём равен 256.

???коэффициент надежности :   1 - 2alpha = 0.95   alpha = 0.025 ?? - нашла в интернете и непонтно к чему это выражение, автор его пото не использует в формуле
по таблице значений функции Лапласа http://www.mathprofi.ru/files/u/laplas.png 


import math

xb = 80
volume = 256
nadegnost = 0.95
sygma = 16
t = nadegnost / 2
80 - t * 16/math.sqrt(256) < a < 80 + t * 16/math.sqrt(256)

In [3]:
import math

xb = 80
volume = 256
nadegnost = 0.95
sygma = 16
t = nadegnost / 2

a1 = 80 - t * 16/math.sqrt(256)
a1

79.525

In [4]:
a2 = 80 + t * 16/math.sqrt(256)
a2

80.475

Доверительный интервал 79.525 < a < 80.475

Задача 2

Продавец утверждает, что средний вес пачки печенья составляет 200 г. Из партии извлечена выборка из 10 пачек. Вес каждой пачки составляет:

202, 203, 199, 197, 195, 201, 200, 204, 194, 190

Известно, что их веса распределены нормально.
1. Верно ли утверждение продавца, если учитывать, что уровень значимости alpha равен 1% == 0.01?
2. Найдите P-значение для данного теста.

H0 = 200 и H1 != 200

In [33]:
import numpy as np
import statistics 

#list = np.array([202, 203, 199, 197, 195, 201, 200, 204, 194, 190])
list = (202, 203, 199, 197, 195, 201, 200, 204, 194, 190)
x_aver = sum(list) / len(list)
x_aver
#list.mean()

198.5

Выборочное среднеквадратичное отклонение:

In [34]:
alpha = 0.01
s = np.var(list)
s

17.85

In [35]:
var_res = sum((i - x_aver) ** 2 for i in list) / len(list)
var_res

17.85

t-распределение. При заданной вероятности рассчитаем предельное отклонение от средней арифметической - границы доверительного интервала:

In [36]:
t1 = x_aver - s / math.sqrt(sum(list)) * alpha
t1

198.49599356625725

In [37]:
t2 = x_aver + s / math.sqrt(sum(list)) * alpha
t2

198.50400643374275

Z-статистика

In [44]:
from scipy import stats

scipy.stats.zscore(list, axis=0, ddof=0)

array([ 0.82841687,  1.0651074 ,  0.11834527, -0.3550358 , -0.82841687,
        0.59172634,  0.3550358 ,  1.30179794, -1.0651074 , -2.01186954])

https://habr.com/ru/post/557424/   
В чем бы не измерялись сравниваемые величины, после вычисления Z-значений они будут измеряться в сигмах. Чем меньше по модулю Z-значение, тем ближе оно к вершине распределения, а знак укажет по какую сторону от вершины это значение находится. Z-значение величины оценивает вероятность (частоту) ее появления

p-value показывает вероятность случайного выхода за границы интервала, задаваемого значением Z-статистики   x: The t-score df: The degrees of freedom     https://habr.com/ru/company/skillfactory/blog/510688/

In [ ]:
scipy.stats.t.sf(abs(x), df)

In [52]:
stat, p = scipy.stats.normaltest(list) # Критерий согласия Пирсона
print('Statistics=%.3f, p-value=%.3f' % (stat, p))
alpha = 0.01
if p > alpha:
    print('Принять гипотезу о нормальности')
else:
    print('Отклонить гипотезу о нормальности')

Statistics=1.027, p-value=0.598
Принять гипотезу о нормальности


Задача 3

Утверждается, что шарики для подшипников, изготовленные автоматическим станком, имеют средний диаметр 17 мм. Используя уровень значимости 0.025 и правостороннюю критическую область, проверить данную гипотезу при условии, что в выборке из 100 шариков средний диаметр оказался равным 17.5 мм, а дисперсия известна и равна 4 мм.

H0 = 10
H1 > 10

In [56]:
x_aver0 = 17
x_aver1 = 17.5
D = 4
alpha = 0.025
n = 100

In [63]:
#фактическое наблюдение
U = ((x_aver1 - x_aver0) / D) * math.sqrt(n)
U

1.25

In [64]:
Laplas_functio = (1 - 2 * alpha) / 2
Laplas_functio

0.475

В таблице лапласа https://100task.ru/sample/119.aspx найдем критическую точку для односторонней критической 
области 0.475 = 1.96

1.25 < 1.96 поэтому нулевую гипотезу можно принять

Задача 4 (Дополнительно)

Дана выборка диаметров подшипников из примера 1 с занятия 5:

samples = [0.6603, 0.9466, 0.5968, 1.3792, 1.5481, 0.7515, 1.0681, 1.1134,
1.2088, 1.701 , 1.0282, 1.3579, 1.0191, 1.1784, 1.1168, 1.1372,
0.7273, 1.3958, 0.8665, 1.5112, 1.161 , 1.0232, 1.0865, 1.02 ]

Предполагая, что диаметры подшипников распределены нормально, проверьте гипотезу о том, что дисперсия случайной величины равна 0.0625 при уровне значимости alpha = 0.05. Что для этого нужно знать:
1. Альтернативная гипотеза двухсторонняя.
2. Статистика для теста: H = (n - 1) * sample_variance / variance, где n - число элементов в выборке, sample_variance - несмещённая оценка дисперсии, variance - утверждаемая нулевой гипотезой дисперсия.
3. Эта статистика в предположении верности нулевой гипотезы имеет распределение хи-квадрат с параметром df = n - 1. Её квантили можно найти с помощью функции scipy.stats.chi2.ppf.